### PR TITLE
lost QoS>=1 messages if message is enqueued in session termination window

### PR DESCRIPTION
**Pull request by dergraf**

### CODE REVIEW

**Comment by dergraf** [comment]
@larshesel please review, cc: @ioolkos

**Comment by larshesel** [comment]
Didn't you add a test for this bug? I don't see it.

File: test/vmq_queue_SUITE.erl [hunk 1]
```diff
@@ -176,6 +178,28 @@ queue_lifo_offline_drop_test(_) ->
...
+    Parent = self(),
+    SubscriberId = {"", <<"mock-lifo-client">>},
+    QueueOpts = maps:merge(vmq_queue:default_opts(), #{clean_session => false,
+                                                       max_offline_messages => 1000,
+                                                       queue_type => fifo}),
+    SessionPid1 = spawn(fun() -> mock_session(Parent) end),
+    {ok, false, QPid} = vmq_reg_leader:register_subscriber(SessionPid1, SubscriberId, QueueOpts),
+    ok = vmq_reg:subscribe(false, <<"mock-user">>, SubscriberId, [{[<<"test">>, <<"transition">>], 1}]),
+    %% teardown session
+    timer:sleep(10), % give some time to plumtree
```
Comment (larshesel) [inline test/vmq_queue_SUITE.erl:190]: Is there not a way we can detect if plumtree has reached a consistent state?

Comment (dergraf) [inline test/vmq_queue_SUITE.erl:190]: currently not, let's follow up on this over #35

File: test/vmq_queue_SUITE.erl [hunk 2]
```diff
@@ -176,6 +178,28 @@ queue_lifo_offline_drop_test(_) ->
...
+    SubscriberId = {"", <<"mock-trans-client">>},
+    QueueOpts = maps:merge(vmq_queue:default_opts(), #{clean_session => false,
+                                                       max_offline_messages => 1000,
+                                                       queue_type => fifo}),
+    SessionPid1 = spawn(fun() -> mock_session(Parent) end),
+    {ok, false, QPid} = vmq_reg_leader:register_subscriber(SessionPid1, SubscriberId, QueueOpts),
+    ok = vmq_reg:subscribe(false, <<"mock-user">>, SubscriberId, [{[<<"test">>, <<"transition">>], 1}]),
+    %% teardown session
+    timer:sleep(10), % give some time to plumtree
+    catch vmq_queue:set_last_waiting_acks(QPid, []), % simulate what real session does
```
Comment (larshesel) [inline test/vmq_queue_SUITE.erl:193]: Why do you need to catch here?

File: test/vmq_queue_SUITE.erl [hunk 3]
```diff
@@ -176,6 +178,28 @@ queue_lifo_offline_drop_test(_) ->
...
+    SessionPid1 ! {go_down_in, 1},
+    Msgs = publish_multi([<<"test">>, <<"transition">>]), % publish 100
```
Comment (larshesel) [inline test/vmq_queue_SUITE.erl:195]: Why do you need to publish 100 messages?

File: test/vmq_queue_SUITE.erl [hunk 4]
```diff
@@ -176,6 +178,28 @@ queue_lifo_offline_drop_test(_) ->
...
+queue_offline_transition_test(_) ->
+    Parent = self(),
+    SubscriberId = {"", <<"mock-trans-client">>},
+    QueueOpts = maps:merge(vmq_queue:default_opts(), #{clean_session => false,
+                                                       max_offline_messages => 1000,
+                                                       queue_type => fifo}),
+    SessionPid1 = spawn(fun() -> mock_session(Parent) end),
+    {ok, false, QPid} = vmq_reg_leader:register_subscriber(SessionPid1, SubscriberId, QueueOpts),
+    ok = vmq_reg:subscribe(false, <<"mock-user">>, SubscriberId, [{[<<"test">>, <<"transition">>], 1}]),
+    %% teardown session
```
Comment (larshesel) [inline test/vmq_queue_SUITE.erl]: This comment seem to be malplaced? Where does the teardown start?

File: test/vmq_queue_SUITE.erl [hunk 5]
```diff
@@ -176,6 +178,28 @@ queue_lifo_offline_drop_test(_) ->
...
+    timer:sleep(10), % give some time to plumtree
+    catch vmq_queue:set_last_waiting_acks(QPid, []), % simulate what real session does
+    SessionPid1 ! {go_down_in, 1},
+    Msgs = publish_multi([<<"test">>, <<"transition">>]), % publish 100
+
+    SessionPid2 = spawn(fun() -> mock_session(Parent) end),
+    {ok, true, QPid} = vmq_reg_leader:register_subscriber(SessionPid2, SubscriberId, QueueOpts),
+    ok = receive_multi(QPid, 1, Msgs),
+    timer:sleep(10),
```
Comment (larshesel) [inline test/vmq_queue_SUITE.erl]: What does this sleep do?

**Comment by larshesel** [comment]
The change looks good - I think the test is harder to understand, but there I also see a lot of things (functions) used I haven't seen before, so sorry for all the Qs.

One other note regarding the tests, the function receive_multi seem to only check that it receives the correct number of messages, would it be possible to strengthen it to also match on the messages?

**Pull request merged**